File: tdd-contract-review/benchmark/sample-app/app/models/transaction.rb
# frozen_string_literal: true

class Transaction < ApplicationRecord
  belongs_to :user
  belongs_to :wallet

  enum :status, { pending: 'pending', completed: 'completed', failed: 'failed', reversed: 'reversed' }
  enum :category, {
    transfer: 'transfer',
    payment: 'payment',
    deposit: 'deposit',
    withdrawal: 'withdrawal'
  }

  validates :amount, presence: true, numericality: { greater_than: 0, less_than_or_equal_to: 1_000_000 }
  validates :currency, presence: true, inclusion: { in: %w[USD EUR GBP BTC ETH] }
  validates :status, presence: true
  validates :description, length: { maximum: 500 }

  scope :recent, -> { where('created_at > ?', 30.days.ago) }
  scope :by_status, ->(status) { where(status: status) }

  after_create :notify_payment_gateway, if: :payment?

  private

  def notify_payment_gateway
    PaymentGateway.charge(
      amount: amount,
      currency: currency,
      user_id: user_id,
      transaction_id: id
    )
  end
end

File: tdd-contract-review/benchmark/sample-app-ts/src/models/Transaction.ts
export type Currency = "USD" | "EUR" | "JPY" | "GBP";

export interface Transaction {
  id: string;
  userId: string;
  amount: number;
  currency: Currency;
  createdAt: string;
}

export function isSupportedCurrency(c: string): c is Currency {
  return c === "USD" || c === "EUR" || c === "JPY" || c === "GBP";
}

export function formatAmount(tx: Transaction): string {
  return `${tx.currency} ${tx.amount.toFixed(2)}`;
}
